internal/util/retry/retry.go

```diff
@@ -78,3 +83,54 @@ func IsUnRecoverable(err error) bool {
 	_, isUnrecoverable := err.(unrecoverableError)
 	return isUnrecoverable
 }
+
+func IsRetryErrorCode(code commonpb.ErrorCode) bool {
+	return code == commonpb.ErrorCode_NotReadyServe ||
+		code == commonpb.ErrorCode_NotShardLeader ||
+		code == commonpb.ErrorCode_NodeIDNotMatch
+}
+
+func DoGrpc(times uint, rpcFunc func() (any, error)) (any, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	var (
+		result     any
+		innerError error
+	)
+
+	err := Do(ctx, func() error {
+		result, innerError = rpcFunc()
+		if innerError != nil {
+			result = nil
+			cancel()
+			return innerError
+		}
+
+		var errorCode commonpb.ErrorCode
+		var reason string
+		switch res := result.(type) {
+		case *commonpb.Status:
+			errorCode = res.GetErrorCode()
+			reason = res.GetReason()
+		case interface{ GetStatus() *commonpb.Status }:
+			errorCode = res.GetStatus().GetErrorCode()
+			reason = res.GetStatus().GetReason()
+		default:
+			cancel()
+			return innerError
+		}
+
+		if errorCode == commonpb.ErrorCode_Success {
+			return nil
+		}
+		innerError = errors.New(reason)
+		if errorCode != commonpb.ErrorCode_NotReadyServe {
+			cancel()
+		}
+		return innerError
+	}, Attempts(times), Sleep(time.Second))
```

Comment:
Sleep is too long. 
1. do we consider about context? what is search timeout is only 100ms?
2. retry.Do use exponential backoff,  reasonable timeout should be something like 50ms, 100ms, 200ms, 400ms, 800ms, 1s, 1s....